### PR TITLE
Adds Indexer state to GRPC health checks [DPP-434]

### PR DESCRIPTION
**Pull request by tudor-da**
Adds participant Indexer state to GRPC health checks.

CHANGELOG_BEGIN
[Integration Kit] The state of the participant indexer can now be checked via the GRPC health endpoint
CHANGELOG_END

### Pull Request Checklist

- [x] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/main/CONTRIBUTING.md)
- [x] Include appropriate tests
- [x] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

**Comment by SamirTalwar-DA** [comment]
I think this will take care of #7237, finally.

**Review by SamirTalwar-DA — APPROVED** [review]
Looks great.

**Review by kamil-da — APPROVED** [review]

File: ledger/participant-integration-api/src/test/suite/scala/platform/indexer/RecoveringIndexerSpec.scala [hunk 1]
```diff
@@ -147,41 +151,96 @@ final class RecoveringIndexerSpec
...
+
+    "report correct health status updates on failures and recoveries" in {
+      val healthStatusLogCapture = mutable.ArrayBuffer.empty[HealthStatus]
+      val healthStatusRef = new AtomicReference[HealthStatus]()
+
+      val recoveringIndexer =
+        new RecoveringIndexer(
+          actorSystem.scheduler,
+          actorSystem.dispatcher,
+          10.millis,
```
Comment (miklos-da) [inline ledger/participant-integration-api/src/test/suite/scala/platform/indexer/RecoveringIndexerSpec.scala:195]: You might want to specify this as a constant e.g. `someTimeout` so you don't have to repeat it 10+ times in this file.

File: ledger/participant-integration-api/src/test/suite/scala/platform/indexer/RecoveringIndexerSpec.scala [hunk 2]
```diff
@@ -101,43 +105,43 @@ final class RecoveringIndexerSpec
...
+      } yield {
+        readLog() should contain theSameElementsInOrderAs Seq(
+          Level.INFO -> "Starting Indexer Server",
+          Level.INFO -> "Started Indexer Server",
+          Level.INFO -> "Successfully finished processing state updates",
+          Level.INFO -> "Stopping Indexer Server",
+          Level.INFO -> "Stopped Indexer Server",
+        )
+        testIndexer.openSubscriptions shouldBe mutable.Set.empty
+        healthReporter.currentHealth() shouldBe HealthStatus.healthy
```
Comment (miklos-da) [inline ledger/participant-integration-api/src/test/suite/scala/platform/indexer/RecoveringIndexerSpec.scala:138]: Does this imply that a stopped indexer server is still healthy?

File: ledger/participant-integration-api/src/main/scala/platform/indexer/RecoveringIndexer.scala [hunk 3]
```diff
@@ -142,4 +147,27 @@ private[indexer] final class RecoveringIndexer(
         })
     })
   }
+
+  private def reportErrorState(errMsg: String, exception: Throwable): Unit = {
```
Comment (miklos-da) [inline ledger/participant-integration-api/src/main/scala/platform/indexer/RecoveringIndexer.scala:151]: Nit:
```suggestion
  private def reportErrorState(errorMessage: String, exception: Throwable): Unit = {
```

File: ledger/participant-integration-api/src/test/suite/scala/platform/indexer/RecoveringIndexerSpec.scala [hunk 4]
```diff
@@ -147,41 +151,96 @@ final class RecoveringIndexerSpec
...
+    "report correct health status updates on failures and recoveries" in {
+      val healthStatusLogCapture = mutable.ArrayBuffer.empty[HealthStatus]
+      val healthStatusRef = new AtomicReference[HealthStatus]()
+
+      val recoveringIndexer =
+        new RecoveringIndexer(
+          actorSystem.scheduler,
+          actorSystem.dispatcher,
+          10.millis,
+          updateHealthStatus = hs => {
```
Comment (miklos-da) [inline ledger/participant-integration-api/src/test/suite/scala/platform/indexer/RecoveringIndexerSpec.scala:196]: Nit:
```suggestion
          updateHealthStatus = healthStatus => {
```

**Review by miklos-da — APPROVED** [review]

File: ledger/participant-integration-api/src/test/suite/scala/platform/indexer/RecoveringIndexerSpec.scala (hunk 2 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/indexer/RecoveringIndexerSpec.scala:138]: If it shut down gracefully, yes.. Do you think it should be non-healthy?

Comment (miklos-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/indexer/RecoveringIndexerSpec.scala:138]: What does the client expect? E.g., can it expect the service that propagates this health status to be still working or not?

File: ledger/participant-integration-api/src/test/suite/scala/platform/indexer/RecoveringIndexerSpec.scala (hunk 4 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/indexer/RecoveringIndexerSpec.scala]: Done

File: ledger/participant-integration-api/src/main/scala/platform/indexer/RecoveringIndexer.scala (hunk 3 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/main/scala/platform/indexer/RecoveringIndexer.scala]: Done

File: ledger/participant-integration-api/src/test/suite/scala/platform/indexer/RecoveringIndexerSpec.scala (hunk 1 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/indexer/RecoveringIndexerSpec.scala]: Done

File: ledger/participant-integration-api/src/test/suite/scala/platform/indexer/RecoveringIndexerSpec.scala (hunk 2 above)
Comment (tudor-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/indexer/RecoveringIndexerSpec.scala:139]: I'm not sold on either message:
If the service is stopped, it's weird to argue that it is healthy. On the other hand, it is weird to also say that it is unhealthy.

Comment (tudor-da) [reply ledger/participant-integration-api/src/test/suite/scala/platform/indexer/RecoveringIndexerSpec.scala:139]: Ah, I think this settles it. This is what grpcurl prints for unhealthy/healthy:
```
{
  "status": "NOT_SERVING"
}
{
  "status": "SERVING"
}
```
So we will go with: unhealthy if the service has stopped.

**Pull request merged**